Interactive Currency Calculations

vernak2539 opened the pull request:
This PR introduces the ability to do to currency exchange rates interactively via input boxes. I've put a video below of the functionality.


https://user-images.githubusercontent.com/521270/163688870-fc583877-f5b6-41f2-8127-00e811d0e36e.mov


Open questions:
1. Should I add tests (3/4 tests of this package fail on main 🤷 )? Not sure how to actually do that based on the code living in the `index.js` page which is then inserted into the `window` on the HTML page
2. I used the `convert` function from `services.js` by inserting it into the client code. Wasn't sure there was another way to do this. Can duplicate the code if needed.

vernak2539 commented:
@jejopl have you had a chance to look at this?

jejopl commented:
Hey @vernak2539, I didn't, sorry! Was super busy during mainnet launch. I'll be on vacation until 25th of July.
I will check it out when I get back. If not, please tag me again, thanks!

vernak2539 commented:
hey @jejopl just tagging you again. let me know if there's anything I can do

jejopl commented:
Hey @vernak2539, sorry for the delay. Looks good, great job!
Can you check "allow edit" box? I wanted to push some small changes to this PR. 

Also there will be one more change that will be cool to have. I'll explain it later, when I will be able to push the code to this PR.

Cheers!

vernak2539 commented:
>Can you check "allow edit" box?

Done!

jejopl commented:
So, I've pushed some small changes. Basically it was mostly about display - the high numbers were pretty hard to read. 
There's one more feature that it would be cool to have. Are you good with adding up/down arrows back? It will require some more work, since you will have to implement it manually. 
You can get the icons from fontawesome, just copy the svg code.
https://fontawesome.com/search?q=chevron&o=r&m=free

Let me know if this is something that you are willing to do.

vernak2539 commented:
I can have a go at adding up and down arrows back yep. Will likely take me a bit to get to it being honest.

Question about the usage of `en-us`. Are we good defaulting to that?

jejopl commented:
Hmm, good catch. We can use client browser setting for this

vernak2539 commented on packages/currencyConverter/index.js:
```diff
@@ -2,28 +2,37 @@
 const {parseAndNormalize, fetchRates, convert} = require('./services');
 
 async function currencyConverter(query, API_KEY) {
+  const browserLang = "en-US"; // how to get dynamically as this is not executed in browser?
```
@jejopl any ideas on this one? Since this part isn't executed in the browser, any place you could point me to? Or do you think that I should init the values on the browser side with a "flash"?

jejopl replied:
if you dont pass any argument to `toLocaleString` it will use the browser locale. So you can just use `toLocaleString()` instead of `toLocaleString("en-US")`

vernak2539 commented:
Hey hey. @threej-in happy to work on this together! Which PR would you like to work off of?

threej-in commented:
Hello @vernak2539, nice to meet you.

As you can see in my PR #156 I also want to have interactive input fields for live conversion on the front end when conversion rates are requested. But you have implemented it better than I because in my code I haven't implemented localization and haven't checked for negative inputs also. So I think we should keep working on your PR.

If you allow me, I would like to add some styling to input fields and also add other features which I have implemented and specified in detail in my PR #156. Please have a look into it and let me know what features should be implemented and which ones to ignore.

vernak2539 commented:
@threej-in sounds good to me. I'll make you a collaborator on my repo in a moment.

As for the features in your PR that can come over, please feel free to do what you see fit. I put my thoughts below on the things you mentioned in your descriptions
1. Fixed unnecessary calls to parseAnsNormalize function after query is triggered by storing the result to global variable (yes, please!)
2. Fixed unnecessary call to crypto API when only fiat conversion is requested. It will help reducing no of calls to API (yes, please!)
3. Fixed problem with thousand separator. Result shown in the below image. (I think this may be fixed by localisation in this branch?)
4. Using Google finance API for fiat conversion to get more accurate data (sure! is there a cost to this? [asking as I don't know])
5. Fixed problems with rounding conversions. Previously huge difference in rate resulted into 0 as toValue. For example: IRR to USD conversion returned 0. Result shown in the below image. (I think this is fixed in this PR?)

I think the only thing that I was hoping to do further in this PR is to add the increment and decrement buttons to the UI (as if we were using an input with a type of "number"). @jejopl discussed this [here](https://github.com/PresearchOfficial/presearch-packages/pull/146#issuecomment-1240707631)

threej-in commented:
> 3. Fixed problem with thousand separator. Result shown in the below image. (I think this may be fixed by localisation in this branch?)

Yes you have already solved this, I forgot to mention.

> 4. Using Google finance API for fiat conversion to get more accurate data (sure! is there a cost to this? [asking as I don't know])

Yes It's free as of now.

> 5. Fixed problems with rounding conversions. Previously huge difference in rate resulted into 0 as toValue. For example: IRR to USD conversion returned 0. Result shown in the below image. (I think this is fixed in this PR?)

Yes it only works when we input the values near to conversion rate, because the resultant value is limited to show 2 decimal places only. So if you search for irr to usd it shows $0 instead of [$0.000024]

 
> I think the only thing that I was hoping to do further in this PR is to add the increment and decrement buttons to the UI (as if we were using an input with a type of "number"). @jejopl discussed this [here](https://github.com/PresearchOfficial/presearch-packages/pull/146#issuecomment-1240707631)

Curious to know, why not to implement the native HTML5 arrows? By giving the input fields as type=number attribute.

vernak2539 commented:
>Yes it only works when we input the values near to conversion rate, because the resultant value is limited to show 2 decimal places only. So if you search for irr to usd it shows $0 instead of [$0.000024]

This is a great point. Would be good to see the actual value if it's well outside two decimal points

>Curious to know, why not to implement the native HTML5 arrows? By giving the input fields as type=number attribute.

I believe this was due to the locally formatted currency and how it can now include commas, which is not technically 
part of a number.

threej-in commented:
@vernak2539 @jejopl please review the changes.

vernak2539 commented:
Hey @threej-in, I'll have a look at them soon. Sorry it's taken me so long, but came down with covid and just getting recovered now

threej-in commented:
> Hey @threej-in, I'll have a look at them soon. Sorry it's taken me so long, but came down with covid and just getting recovered now

No worries... Please take care of your health.

vernak2539 commented on packages/currencyConverter/services.js:
```diff
@@ -7,6 +7,7 @@ const fiatCurrenciesExtended = require("./fiatCurrenciesExtended.json");
 
 const FIAT_API = "https://ec.europa.eu/budg/inforeuro/api/public/monthly-rates";
 const CRYPTO_API = "https://pro-api.coinmarketcap.com/v1/cryptocurrency/quotes/latest";
+const SPREADSHEET_URI = "https://docs.google.com/spreadsheets/d/e/2PACX-1vRfp_xd8g5Sgw3v2rfgOSbMPnNaSkZsV6KMjJRPFDxY5Kb9AJoyUv8v54IqK-bHgjFV4efHrXntXGf5/pub?output=csv";
```
Just wondering, who owns this spreadsheet? Is there a reason not to embed it? Thinking about a potential security concern

vernak2539 commented:
I've made some small changes and popped in a comment. Other than that, @jejopl it looks like it may be time for you to have a look.

jejopl commented:
Thanks @vernak2539, @threej-in I will take a look

jejopl commented:
it looks like that something is wrong with the calculation, when you are changing the input

https://user-images.githubusercontent.com/32012330/193252196-0ad8ac9c-7032-4de6-9bc4-6f38840d6591.mov

threej-in commented:
> it looks like that something is wrong with the calculation, when you are changing the input
Fixed

vernak2539 commented:
@jejopl any update on this by chance?

jejopl commented:
Hey @vernak2539, sorry I was pretty busy lately. Will try to review it ASAP and I will get back to you guys. 
Appreciate your patience 🙏

jejopl commented:
@vernak2539 I've found one more issue. When on the initial search, when you are using arrows, the currency calculation is wrong. Also it would be cool to auto select the whole input on click

https://user-images.githubusercontent.com/32012330/195081361-22469ab2-e9c0-4d60-9d6b-a6d1ff7b4ce3.mov

threej-in commented:
> @vernak2539 I've found one more issue. When on the initial search, when you are using arrows, the currency calculation is wrong. Also it would be cool to auto select the whole input on click

Solved, please verify.

jejopl commented:
thanks @threej-in, I've found some additional issues. 
The first currency is not displayed correctly. 
Instead of 100$ there's 10$

Also, when you click on the input and then click away, there's "NaN" value

https://user-images.githubusercontent.com/32012330/195331914-5f6d4fbf-81a9-473e-897a-2fd2fe0a9006.mov

threej-in commented:
I believe that the bug is completely fixed now. Let me know if any other changes are required.

vernak2539 commented:
@threej-in it looks like there are some merge conflicts. would you like me to handle them or would you like to do so?

jejopl commented:
Hey @vernak2539, need to check this again - sorry for the delay. There were some changes on the backend about how we are handling the packages and I had to finish that first. I'll review it shortly and if it's good - it will be merged

jejopl commented:
@vernak2539 @threej-in sorry for late review. I have one issue when I search for `100 usd to eur` 

![image](https://user-images.githubusercontent.com/32012330/205120379-d9c3ecec-1981-47e0-9a4d-1b64400a2654.png)

threej-in commented:
Hello @jejopl, looks like it is a platform related issue or something else, I am not getting anything like this, everything's working fine on my machine. Can you please send me the exact steps to recreate the error.

jejopl commented:
@threej-in I found the issue. The problem was that `toLocaleString()` has been execuded in node.js and not on the client side. Because of that, based on the node location the output will be different. It was setting `toValue` to the number with `,` as separator

![image](https://user-images.githubusercontent.com/32012330/205271129-489d0dba-9e20-485b-86b7-720bca28f852.png)

Great job guys, it's working very well! 😃 👍

vernak2539 commented:
Thanks @threej-in for all the work and help with this! Thanks @jejopl for getting this through!
